src/controller/java/src/chip/devicecontroller/NsdManagerServiceResolver.java

```diff
@@ -0,0 +1,73 @@
+/*
+ *   Copyright (c) 2021 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+package chip.devicecontroller;
+
+import android.content.Context;
+import android.net.nsd.NsdManager;
+import android.net.nsd.NsdServiceInfo;
+import android.util.Log;
+
+public class NsdManagerServiceResolver implements ServiceResolver {
+  private final String TAG = NsdManagerServiceResolver.class.getSimpleName();
+  private final NsdManager nsdManager;
+
+  public NsdManagerServiceResolver(Context context) {
+    this.nsdManager = (NsdManager) context.getSystemService(Context.NSD_SERVICE);
+  }
+
+  @Override
+  public void resolve(
+      final String instanceName,
+      final String serviceType,
+      final long callbackHandle,
+      final long contextHandle) {
+    NsdServiceInfo serviceInfo = new NsdServiceInfo();
+    serviceInfo.setServiceName(instanceName);
+    serviceInfo.setServiceType(serviceType);
+
+    this.nsdManager.resolveService(
+        serviceInfo,
+        new NsdManager.ResolveListener() {
+          @Override
+          public void onResolveFailed(NsdServiceInfo serviceInfo, int errorCode) {
+            Log.w(
+                TAG,
+                "Failed to resolve service '" + serviceInfo.getServiceName() + "': " + errorCode);
+            ChipDeviceController.handleServiceResolve(
+                instanceName, serviceType, null, 0, callbackHandle, contextHandle);
+          }
+
+          @Override
+          public void onServiceResolved(NsdServiceInfo serviceInfo) {
+            Log.i(
+                TAG,
+                "Resolved service '"
+                    + serviceInfo.getServiceName()
+                    + "' to "
+                    + serviceInfo.getHost());
+            ChipDeviceController.handleServiceResolve(
```

Comment:
Hmm.  I would have expected a NetworkInterface somewhere here so link-local ipv6 works....  How does that end up working?

Reply:
I don't think Android exposes the interface id through `NsdServiceInfo`: https://developer.android.com/reference/android/net/nsd/NsdServiceInfo. I verified that I can ping other devices in my home network from my phone even I have both Wi-Fi and LTE enabled, but I'm not sure if it's because Wi-Fi is the default route or there's some other magic in Android. Do you think I could file an issue to investigate that as follow up?

Reply:
In theory, we could iterate over all interface and test them with ICMP whether a given address is reachable, but that would introduce some delay, so I would prefer to know for sure that we have no other choice.

Reply:
Followup issue to sort out what happens here for link-local ipv6 is definitely just fine.  If it just works, great.  :)